### PR TITLE
o/snapstate: implement download-component handler

### PR DESCRIPTION
**Pull request by andrewphelpsj**
Implement a task handler for downloading components from the store. Will be used in subsequent PRs for installing components and snaps from the store at the same time.

### CODE REVIEW

File: overlord/snapstate/handlers_components.go [hunk 1]
```diff
@@ -129,6 +130,82 @@ func (m *SnapManager) doPrepareComponent(t *state.Task, _ *tomb.Tomb) error {
...
+
+	cpi := snap.MinimalComponentContainerPlaceInfo(
+		compsup.ComponentName(), compsup.CompSideInfo.Revision,
+		snapsup.InstanceName(),
+	)
+
+	// TODO: to be consistent with snaps, this should be able to point somewhere
+	// else, based on a path that is in the compsup. this would be used for
+	// creating new recovery systems, like it is now for snaps
+	target := cpi.MountFile()
```
Comment (alfonsosanchezbeato) [inline overlord/snapstate/handlers_components.go:174]: It should be possible to re-use `snapsup.DownloadBlobDir` for this and implement something similar to `SnapSetup.MountFile()` but for components I guess.

File: overlord/snapstate/handlers_components.go [hunk 2]
```diff
@@ -129,6 +130,82 @@ func (m *SnapManager) doPrepareComponent(t *state.Task, _ *tomb.Tomb) error {
...
+			user,
+			&store.DownloadOptions{
+				Scheduled: snapsup.IsAutoRefresh,
+				RateLimit: rate,
+			},
+		)
+	})
+	st.Lock()
+	if err != nil {
+		return fmt.Errorf("cannot download component %q: %w", compsup.ComponentName(), err)
```
Comment (alfonsosanchezbeato) [inline overlord/snapstate/handlers_components.go:197]: This path is not tested

**Review by alfonsosanchezbeato — COMMENTED** [review]
Thanks! Some comments below

File: overlord/snapstate/handlers_components.go (hunk 1 above)
Comment (andrewphelpsj) [reply overlord/snapstate/handlers_components.go:174]: Yeah, that would work. I will handle this in a later PR.

**Review by alfonsosanchezbeato — APPROVED** [review]
Thanks for the changes, LGTM

File: overlord/snapstate/backend_test.go [hunk 3]
```diff
@@ -181,9 +181,20 @@ type fakeStore struct {
 	state           *state.State
 	seenPrivacyKeys map[string]bool
 
+	// snapResources is called for each snap that gets returned by SnapAction,
```
Comment (MiguelPires) [inline overlord/snapstate/backend_test.go:184]:
```suggestion
	// snapResourcesFn is called for each snap that gets returned by SnapAction,
```

File: overlord/snapstate/handlers_components.go [hunk 4]
```diff
@@ -129,6 +130,82 @@ func (m *SnapManager) doPrepareComponent(t *state.Task, _ *tomb.Tomb) error {
...
+			compsup.CompSideInfo.Component.String(),
+			target,
+			compsup.DownloadInfo,
+			meter,
+			user,
+			&store.DownloadOptions{
+				Scheduled: snapsup.IsAutoRefresh,
+				RateLimit: rate,
+			},
+		)
```
Comment (MiguelPires) [inline overlord/snapstate/handlers_components.go:193]: bit of a nitpick but we could pull out the component ref and download options and then make this one line. IMO, it's a bit easier to read (it's subjective, so feel free to keep it if you like)
```suggestion
		err = sto.Download(tomb.Context(nil), compRef, target, compsup.DownloadInfo, meter, user, dlOpts)
```

File: overlord/snapstate/handlers_components_download_test.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+	si := &snap.SideInfo{
+		RealName: "snap",
+		SnapID:   snaptest.AssertedSnapID("snap"),
+		Revision: snap.R(11),
+		Channel:  "latest/stable",
+	}
+
+	t := s.state.NewTask("download-component", "...")
+
+	t.Set("snap-setup", &snapstate.SnapSetup{
```
Comment (MiguelPires) [inline overlord/snapstate/handlers_components_download_test.go:61]: The rate limiting isn't tested, there could be a simple variant of this with `AutoRefresh=true` that checks that

File: overlord/snapstate/handlers_components_download_test.go [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+	. "gopkg.in/check.v1"
+)
+
+type downloadComponentSuite struct {
+	baseHandlerSuite
+
+	fakeStore *fakeStore
+}
+
+var _ = Suite(&downloadComponentSuite{})
```
Comment (MiguelPires) [inline overlord/snapstate/handlers_components_download_test.go:29]: These tests target download-component specifically, is there (or will there be) another test that tests this with a more complete change? That would cover whether DownloadInfo is unexpectedly nil and things like that

**Review by MiguelPires — COMMENTED** [review]
Looks good. Some questions around testing

File: overlord/snapstate/handlers_components_download_test.go (hunk 6 above)
Comment (andrewphelpsj) [reply overlord/snapstate/handlers_components_download_test.go:29]: Yes, more complete tests are coming once we can actually use this handler.

File: overlord/snapstate/handlers_components_download_test.go (hunk 5 above)
Comment (andrewphelpsj) [reply overlord/snapstate/handlers_components_download_test.go:77]: Thanks, added in a4ee50ac0ccb1c6f21300936fd6dbf3402e904d5.

**Review by MiguelPires — APPROVED** [review]
lgtm, thank you

**Pull request merged**